src/start.ts

```diff
@@ -54,265 +54,331 @@ export const handler = async (_event: any = {}): Promise<any> => {
+               *    enable the beacon server to make requests to AirnodeRrp
+               * 2. Request sponsor should then call setUpdatePermissionStatus(keeperSponsorWallet.address, true)
+               *    to allow requester to update beacon
+               */
+
+              const requestSponsorWallet = node.evm.deriveSponsorWallet(
+                airnodeHDNode,
+                requestSponsor
+              );
+
+              // Check to prevent sending the same request for beacon update more than once
+              // by checking if a RequestedBeaconUpdate event was emitted but no matching
+              // UpdatedBeacon event was emitted.
+
+              // 1. Fetch RequestedBeaconUpdate events by beaconId, sponsor and sponsorWallet
+              const requestedBeaconUpdateFilter =
+                rrpBeaconServer.filters.RequestedBeaconUpdate(
+                  beaconId,
+                  requestSponsor,
+                  keeperSponsorWallet.address
+                );
+              const requestedBeaconUpdateEvents =
+                await rrpBeaconServer.queryFilter(
+                  requestedBeaconUpdateFilter,
+                  eventLogMaxBlocks * -1,
+                  currentBlock
+                );
+
+              // 2. Fetch UpdatedBeacon events by beaconId
+              const updatedBeaconFilter =
+                rrpBeaconServer.filters.UpdatedBeacon(beaconId);
+              const updatedBeaconEvents = await rrpBeaconServer.queryFilter(
+                updatedBeaconFilter,
+                eventLogMaxBlocks * -1,
+                currentBlock
+              );
+
+              // 3. Match these events by requestId and unmatched events
+              //    are the ones that are still waiting to be fulfilled
+              const [pendingRequestedBeaconUpdateEvent] =
+                requestedBeaconUpdateEvents.filter(
+                  (rbue) =>
+                    !updatedBeaconEvents.some(
+                      (ub) => rbue.args!["requestId"] === ub.args!["requestId"]
+                    )
+                );
+              if (!isNil(pendingRequestedBeaconUpdateEvent)) {
+                // 4. Check if RequestedBeaconUpdate event is awaiting fulfillment by
+                //    calling AirnodeRrp.requestIsAwaitingFulfillment with requestId
+                //    and check if beacon value is fresh enough and skip if it is
+
+                const requestIsAwaitingFulfillment =
+                  await airnodeRrp.requestIsAwaitingFulfillment(
+                    pendingRequestedBeaconUpdateEvent.args!["requestId"]
+                  );
+                if (requestIsAwaitingFulfillment) {
+                  console.log(
+                    "[INFO]\trequest is awaiting fulfillment. skipping update"
+                  );
+                  continue;
+                }
+              }
+
+              // Fetch current gas fee data
+              const [gasPriceLogs, gasTarget] = await getGasPrice({
```

Comment:
In this scenario you'll retrieve the gas pricing for every request, this will slow things down a lot and also increase the requests to the chain gateway/EVM node.

I suggest you move this up so the result can be re-used. This would be on a per-chain-provider basis, so probably to line ~46.

Reply:
This a good point but I had a discussion with Burak where he suggested that we only do this kind of chain requests whenever we are absolutely sure that we must create a beacon request transaction since most of the times Airkeeper will probably not reach this far down. Meaning that Airkeeper jobs running every minute will probably won't trigger a beacon update therefore no gas price data will be needed. Thoughts?

Reply:
Ah, that's an interesting point.

The solution then would probably be to leave as-is for now but ultimately separate deviation assessment and possible result submission (sounds familiar) - and then cache gas pricing during the submission phase.

Reply:
Right, I'm not sure which option is better at the moment and I'll have to profile it at some point but I'll probably come back to this once Airkeeper is ready and it's time to make improvements.

Reply:
Another thing to consider is that we expect to have a single `keeperSponsorWallet` per trigger job meaning that it's a loop because we'd like Airkeeper not to fail but it is expected that this loop runs only once if properly configured.